### PR TITLE
Add typed diag

### PR DESCRIPTION
**Pull request by mcwitt**
#233 

- [x] diag

**EDIT**: this PR just adds `diag` to `Torch.Typed.Functional`. The comments below the line are resolved.

TODO
- [x] Fix invalid dtype, device combos exposed by tests

----
(outdated)

So far I've only attempted `diag`. I approached this by splitting into 2 branches (`upperDiag` and `lowerDiag`), which return diagonals above and below the main diagonal, respectively (this is so the index can be a natural).

I also considered implementing this as a single function with a `Bool` type parameter, i.e. something like

``` haskell
diag
  :: forall (index :: Nat) (upper :: Bool) (i :: Nat) (j :: Nat) device dtype
   . (KnownNat index, KnownBool upper)
  => Tensor device dtype '[i, j] -- ^ input
  -> Tensor device dtype '[Min i j - index] -- ^ output
diag t = unsafePerformIO $ ATen.cast2 ATen.Managed.tensor_diag_l t index
  where
    absIndex = natValI @index
    index = if boolVal (Proxy @upper) then absIndex else -absIndex
```

but splitting into 2 branches seemed simpler, although I'm happy to revisit this.

### CODE REVIEW

**Comment by tscholak** [comment]
Hi @mcwitt, thanks for having a stab at this! Well done!

I only had a brief look, but it seems you could indeed merge these two functions. It seems that since the output type does not depend on weather its lower or upper half, you don't have to bring this boolean up to the type level. You can instead use an input value instead. I recommend using `Torch.Functional.Tri(Upper, Lower)`.

**Comment by tscholak** [comment]
Another thought: You have implemented the case 2d input -> 1d output.
`torch.diag` also supports 1d input -> 2d output, see https://pytorch.org/docs/master/generated/torch.diag.html.

File: hasktorch/src/Torch/Typed/Functional.hs [hunk 1]
```diff
@@ -1191,6 +1191,10 @@ transpose2D
   -> Tensor device dtype '[j, i] -- ^ output
 transpose2D = transpose @0 @1
 
+type family DiagShape (index :: Nat) (shape :: [Nat]) :: [Nat] where
+  DiagShape i '[n] = '[n + i, n + i] -- 1d -> 2d case
+  DiagShape i '[m, n] = '[Min m n - i] -- 2d -> 1d case
```
Comment (tscholak) [inline hasktorch/src/Torch/Typed/Functional.hs:1196]: great!

File: hasktorch/src/Torch/Typed/Functional.hs [hunk 2]
```diff
@@ -1200,11 +1204,13 @@ transpose2D = transpose @0 @1
...
+  :: forall (index :: Nat) (shape :: [Nat]) (shape' :: [Nat]) device dtype
+   . ( KnownNat index
+     , shape' ~ DiagShape index shape
+     )
   => Tri -- ^ upper or lower
-  -> Tensor device dtype '[i, j] -- ^ input
-  -> Tensor device dtype '[Min i j - index] -- ^ output
+  -> Tensor device dtype shape -- ^ input
+  -> Tensor device dtype shape' -- ^ output
 diag tri t = unsafePerformIO $ ATen.cast2 ATen.Managed.tensor_diag_l t
```
Comment (tscholak) [inline hasktorch/src/Torch/Typed/Functional.hs:1214]: you may want to look into writing a test suite for this function in `Torch.Typed.FunctionalSpec`, see, e.g., https://github.com/hasktorch/hasktorch/pull/409/files#diff-e428518635181a3dbe949a37402aa368

Comment (tscholak) [reply hasktorch/src/Torch/Typed/Functional.hs:1214]: for instance, the type signature claims that this function works for all `device`s and all `dtype`s. you may want to make sure that this is accurate :)

Comment (mcwitt) [reply hasktorch/src/Torch/Typed/Functional.hs:1214]: on it, thanks!

File: hasktorch/src/Torch/Typed/Functional.hs [hunk 3]
```diff
@@ -1191,8 +1191,31 @@ transpose2D
   -> Tensor device dtype '[j, i] -- ^ output
 transpose2D = transpose @0 @1
 
--- diag :: Tensor device dtype shape -> Int -> Tensor device dtype shape
--- diag t index = unsafePerformIO $ (ATen.cast2 ATen.Managed.tensor_diag_l) t index
+type family DiagShape (index :: Nat) (shape :: [Nat]) :: [Nat] where
+  DiagShape i '[n] = '[n + i, n + i] -- 1d -> 2d case
+  DiagShape i '[m, n] = '[Min m n - i] -- 2d -> 1d case
+  DiagShape _ _ =  TypeError (Text "The input must be a matrix or a vector.")
```
Comment (tscholak) [inline hasktorch/src/Torch/Typed/Functional.hs:1197]: that's a great addition and will help the user, thanks!
```suggestion
  DiagShape _ shape =  TypeError (Text "The input must be a matrix or a vector, but it has " :<>: ListLength shape  :<>: " dimensions.")
```
if you can, you may also want to guard against invalid `i`s :)

File: hasktorch/test/Torch/Typed/FunctionalSpec.hs [hunk 4]
```diff
@@ -698,6 +698,18 @@ instance
...
+
+instance
+  ( TensorOptions shape dtype device
+  , TensorOptions shape' dtype device
+  , shape' ~ DiagShape index shape
+  , KnownNat index
+  ) => Apply' DiagSpec ((Proxy index, (Proxy device, (Proxy dtype, Proxy shape))), IO ()) (IO ()) where
+  apply' DiagSpec (_, agg) = agg >> do
+    let t = ones @shape @dtype @device
+    foldMap (\tri -> checkDynamicTensorAttributes $ diag @index tri t) [Upper, Lower]
```
Comment (tscholak) [inline hasktorch/test/Torch/Typed/FunctionalSpec.hs:711]: very good!

File: hasktorch/test/Torch/Typed/FunctionalSpec.hs [hunk 5]
```diff
@@ -981,6 +993,21 @@ spec' device =
           hfoldrM @IO Transpose2DSpec () (hattach cpu   (hproduct allDTypes (Proxy @'[2, 3] :. HNil)))
         Device { deviceType = CUDA, deviceIndex = 0 } ->
           hfoldrM @IO Transpose2DSpec () (hattach cuda0 (hproduct allDTypes (Proxy @'[2, 3] :. HNil)))
+      it "diag" $ do
+        let
+          shapes1 = Proxy @'[0] :. Proxy @'[1] :. Proxy @'[3] :. HNil
+          shapes2 = Proxy @'[2, 3] :. HNil
+          shapes2' = Proxy @'[0, 0] :. Proxy @'[0, 1]  :. Proxy @'[1, 0] :. HNil -- only offset 0 has upper and lower defined
```
Comment (tscholak) [inline hasktorch/test/Torch/Typed/FunctionalSpec.hs:1000]: if certain offsets are not allowed with certain shapes, then you may want to add this to your type family above, e.g. 
```hs
type family DiagShape (index :: Nat) (shape :: [Nat]) :: [Nat] where
   ...
   DiagShape 0 '[0, n] = ... ok!
   DiagShape _ '[0, n] = TypeError "boom!"
   DiagShape 0 '[n, 0] = ... ok!
   DiagShape _ '[n, 0] = TypeError "boom!"
   ...
```

Comment (tscholak) [reply hasktorch/test/Torch/Typed/FunctionalSpec.hs:1000]: or maybe there's a more general rule that holds for all `'[m, n]` :)

Comment (mcwitt) [reply hasktorch/test/Torch/Typed/FunctionalSpec.hs:1037]: Thanks for the tip. I think this should be addressed generally by 4523321.

File: hasktorch/test/Torch/Typed/FunctionalSpec.hs [hunk 6]
```diff
@@ -1017,6 +1031,22 @@ spec' device =
...
+          tris = Proxy @'Upper :. Proxy @'Lower :. HNil
+          indexes = Proxy @0 :. Proxy @1 :. HNil
+          indexes' = Proxy @0 :. HNil
+        case device of
+          Device { deviceType = CPU,  deviceIndex = 0 } -> do
+            hfoldrM @IO DiagSpec () (hproduct (hproduct tris indexes)  (hattach cpu   (hproduct standardDTypes standardShapes)))
+            hfoldrM @IO DiagSpec () (hproduct (hproduct tris indexes)  (hattach cpu   (hproduct standardDTypes vectorShapes)))
+            hfoldrM @IO DiagSpec () (hproduct (hproduct tris indexes') (hattach cpu   (hproduct standardDTypes emptyShapes)))
+          Device { deviceType = CUDA, deviceIndex = 0 } -> do
+            hfoldrM @IO DiagSpec () (hproduct (hproduct tris indexes)  (hattach cuda0 (hproduct standardDTypes standardShapes)))
```
Comment (tscholak) [inline hasktorch/test/Torch/Typed/FunctionalSpec.hs:1047]: you may be able to add `Half` precision here

Comment (tscholak) [reply hasktorch/test/Torch/Typed/FunctionalSpec.hs]: thanks!

**Pull request merged**